### PR TITLE
added more security and roles

### PR DESCRIPTION
**Pull request by kriazantseva**

### CODE REVIEW

File: src/main/java/mate/academy/bookstoreapp/controller/BookController.java [hunk 1]
```diff
@@ -49,13 +50,15 @@ public List<BookDto> searchBooks(BookSearchParametersDto searchParametersDto) {
         return bookService.search(searchParametersDto);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```

File: src/main/java/mate/academy/bookstoreapp/controller/BookController.java [hunk 2]
```diff
@@ -49,13 +50,15 @@ public List<BookDto> searchBooks(BookSearchParametersDto searchParametersDto) {
...
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
     @Operation(summary = "Create a book", description = "Create a book")
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto requestDto) {
         return bookService.save(requestDto);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```

File: src/main/java/mate/academy/bookstoreapp/controller/BookController.java [hunk 3]
```diff
@@ -64,6 +67,7 @@ public BookDto updateBook(@PathVariable Long id,
         return bookService.update(id, requestDto);
     }
 
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/controller/BookController.java]: same, spring will add ROLE_ by itself

File: src/main/java/mate/academy/bookstoreapp/model/Role.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    private RoleName role;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + role.name();
+    }
+
+    public enum RoleName {
+        USER,
+        ADMIN
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/model/Role.java]:
```suggestion
        ROLE_USER,
        ROLE_ADMIN

```

Comment (Vasul-Pankiv) [reply src/main/java/mate/academy/bookstoreapp/model/Role.java]: The same changes make in db

File: src/main/java/mate/academy/bookstoreapp/model/Role.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    @Enumerated(EnumType.STRING)
+    private RoleName role;
+
+    @Override
+    public String getAuthority() {
+        return "ROLE_" + role.name();
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/model/Role.java]:
```suggestion
        return role.name();
```
Please read about diff between ROLE and AUTHORITY in Spring

File: src/main/java/mate/academy/bookstoreapp/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -25,6 +35,20 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
 
         User user = userMapper.toUser(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.password()));
+        if (requestDto.email().contains("@admin")) {
+            setRolesToUser(user, Role.RoleName.ADMIN);
+        } else {
+            setRolesToUser(user, Role.RoleName.USER);
+        }
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/service/impl/UserServiceImpl.java]: bad approach and so unsecured
set only USER role while user registration

File: src/main/java/mate/academy/bookstoreapp/validation/PasswordValidation.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package mate.academy.bookstoreapp.validation;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.regex.Pattern;
+
+public class PasswordValidation implements ConstraintValidator<Password, String> {
```
Comment (Vasul-Pankiv) [inline src/main/java/mate/academy/bookstoreapp/validation/PasswordValidation.java]:
```suggestion
public class PasswordValidator implements ConstraintValidator<Password, String> {

```

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

**Review by sarakhmen — APPROVED** [review]
LGTM!

**Pull request merged**